trieste/acquisition/function/greedy.py

```diff
@@ -351,3 +355,127 @@ def __call__(self, x: TensorType) -> TensorType:
+    this acquisition function,
+    TODO: add stuff
+    it provides diverse batches of candidate points.
+    """
+
+    def __init__(
+        self,
+        acquisition_function_builder: AcquisitionFunctionBuilder = None,
+        fantasize_method: str = "KB",
+    ):
+        """
+        :param acquisition_function_builder:
+        :raise tf.errors.InvalidArgumentError: If ``fantasize_method`` is not "KB" or "sample".
+        """
+        if acquisition_function_builder is None:
+            self._builder = ExpectedImprovement()
+        else:
+            self._builder = acquisition_function_builder
+
+        # TODO assert fantasize mode
+        self._fantasize_method = fantasize_method
+
+    def prepare_acquisition_function(
+        self,
+        models: Mapping[str, ProbabilisticModel],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param models: The models over each tag.
+        :param datasets: The data from the observer (optional).
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :return: An acquisition function.
+        """
+        if pending_points is not None:
+            fantasized_data = {
+                tag: _generate_fantasized_data(model, pending_points, self._fantasize_method)
+                for tag, model in models.items()
+            }
+            models = {
+                tag: FantasizedGPRModel(model, fantasized_data[tag])
+                for tag, model in models.items()
+            }
+
+        return self._builder.prepare_acquisition_function(models, datasets)
+
+
+def _generate_fantasized_data(model, pending_points, fantasize_method):
+    if fantasize_method == "KB":
+        fantasized_obs, _ = model.predict(pending_points)
+    elif fantasize_method == "sample":
+        fantasized_obs = model.sample(pending_points, num_samples=1)  # TODO check dim
+    else:
+        raise NotImplementedError(
+            f"Fantasize method not supported. "
+            f"Expected KB or sample, received "
+            f"{fantasize_method}"
+        )
+    return Dataset(pending_points, fantasized_obs)
+
+
+class FantasizedGPRModel(ProbabilisticModel):
+    """
+    Creates new model from an existing GPR model and additional data.
+    This new model posterior is conditioned and the GPR data and the additional one.
+    """
+
+    def __init__(self, model: ProbabilisticModel, fantasized_data: Dataset):
+        """
+
+        :param model: a GPR model
+        :param fantasized_data: additional dataset to condition on
+        :raise NotImplementedError: If model is not of class GaussianProcessRegression.
+        """
+
+        if not isinstance(model, GaussianProcessRegression):
```

Comment:
we don't need to check its GaussianPprocessRegression, since we use FastUpdateModel

we should still check though that conditional methods actually exist so that a meaningful error appears at runtime